inspectkv/inspectkv.go

```diff
@@ -108,20 +123,122 @@ func ScanIndexData(txn kv.Transaction, kvIndex kv.Index, startVals []interface{}
 }
 
-// RecordData is the record data composed of a handle and column values.
-type RecordData struct {
-	Handle int64
-	Values []interface{}
+// ScanIndexColData scans the index handles and values in a limited number, according to the corresponding column.
+// It returns data and the next startHandle until it doesn't have data, then returns data is nil and
+// the next startHandle is the handle which can't get data.
+// If limit = -1, it returns the index data of the whole.
+func ScanIndexColData(retriever kv.Retriever, t table.Table, idx *column.IndexedCol, startHandle, limit int64) (
+	[]*RecordData, int64, error) {
+	cols := make([]*column.Col, len(idx.Columns))
+	for i, col := range idx.Columns {
+		cols[i] = t.Cols()[col.Offset]
+	}
+
+	return scanTableData(retriever, t, cols, startHandle, limit)
 }
 
-// ScanTableData scans table row handles and column values in a limited number.
-// It returns data and the next startKey.
-func ScanTableData(t table.Table, retriever kv.Retriever, startHandle, limit int64) (
+// EqualRecords returns ture if a is equal to b.
+func EqualRecords(a, b []*RecordData) bool {
+	if len(a) != len(b) {
+		return false
+	}
+
+	for i, r := range a {
+		if !r.Equal(b[i]) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// DiffRecords compares records one by one.
+// It returns the difference between a and b.
+// e.g,
+// a {{1, "val0"}, {2, "val2"}}
+// b {{1, "val1"}, {3, "val3"}}
+// returns {{1, "val0"}, {2, "val2"}, nil}, {{1, "val1"}, nil, {3, "val3"}}, nil
+func DiffRecords(a, b []*RecordData) ([]*RecordData, []*RecordData, error) {
+	if len(a) == 0 {
+		return make([]*RecordData, len(b)), b, nil
+	}
+
+	var ret1, ret2 []*RecordData
+	k := 0
+	for i := 0; i < len(a); i++ {
+		if len(b[k:]) == 0 {
+			ret1 = append(ret1, a[i:]...)
+			ret2 = append(ret2, make([]*RecordData, len(a[i:]))...)
+			break
+		}
+
+		for j := k; j < len(b); j++ {
+			if a[i].Handle > b[j].Handle {
+				ret1 = append(ret1, nil)
+				ret2 = append(ret2, b[j])
+				continue
+			}
+			if a[i].Handle < b[j].Handle {
+				ret1 = append(ret1, a[i])
+				ret2 = append(ret2, nil)
+				if len(a[i+1:]) == 0 {
+					ret1 = append(ret1, make([]*RecordData, len(b[j:]))...)
+					ret2 = append(ret2, b[j:]...)
+				}
+				break
+			}
+			if !reflect.DeepEqual(a[i].Values, b[j].Values) {
+				ret1 = append(ret1, a[i])
+				ret2 = append(ret2, b[j])
+			}
+			k = j + 1
+			break
+		}
+	}
+
+	return ret1, ret2, nil
+}
+
+// EqualIndexData returns ture if the data from the index is equal to the data from the table columns.
+func EqualIndexData(txn kv.Transaction, t table.Table, idx *column.IndexedCol) (bool, error) {
+	kvIndex := kv.NewKVIndex(t.IndexPrefix(), idx.Name.L, idx.ID, idx.Unique)
+
+	rs1, _, err := ScanIndexData(txn, kvIndex, nil, -1)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	rs2, _, err := ScanIndexColData(txn, t, idx, 0, -1)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	return EqualRecords(rs1, rs2), nil
```

Comment:
The records returned from `ScanIndexData` has different order than the records retrieved from `ScanIndexColData`, if you want to compare them all, you need to sort `rs1` by handle first.